5_high_order_functions/flattening.js

```diff
@@ -0,0 +1,12 @@
+/*
+ * Use the reduce method in combination with the concat method to “flatten”
+ * an array of arrays into a single array that has all the elements of the original
+ * arrays.
+ */
+
+function flatten(nestedArray) {
+    return nestedArray.reduce((a, b) => a.concat(b));
```

Comment:
Tu función concatena arrays siempre y cuando se trate de arrays de arrays. Es decir, pasar un argumento como este `[1,2,3,[4,5,6]]` va a hacer que tu función truene.

Toma como referencia a la función [`flat`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flat) que ya está implementada en navegadores que si tiene este comportamiento.

Es más, ¿qué pasa si tuvieras múltiples de arrays de arrays anidados? Como esto `[1, 2, 3, [4, 5, 6, [7, 8, 9, [10, 11, 12, [13, 14, 15]]]]]`. ¿Cómo soportarías un caso como ese?

Reply:
Sip, la funcion está hecha solamente para el caso solicitado en el ejercicio. Ya si se quiere que tambien cubra casos de array of numbers and arrays, y de nested arrays, viendo la pagina que pasaste, se agregaría la funcionalidad de un stack en el que primeramente se analizarian los tipos de cada elemento, si es number, pasa directo, y si es array, se hace una funcion recursiva para hacer el primer analisis dentro de ese array e ir extrayendo valores de uno por uno para mandarlos al stack.

Reply:
A lo mejor y no necesitas un stack!

Te mostraré como implementé yo la función "aumentada".

El primer cambio que hice, fue pasar un `[]` vacío como segundo argumento de `reduce()`. El segundo argumento de `reduce()` funcionará como el primer valor a iterar.

Por ejemplo, al final del archivo, cuando mandas llamar `flatten([[1, 2, 3], [4, 5, 6], [7, 8, 9]])`, el primer valor a iterar en la versión anterior era `[1, 2, 3]`, el segundo `[4, 5, 6]`, and so on.

En esta nueva versión, el primer valor a iterar será `[]`, el segundo `[1, 2, 3]`, el tercero `[4, 5, 6]`, etc. 

¿Para qué nos sirve este `[]` inicial?

Nos sirve para capturar los casos donde el primer valor que le pasemos al arreglo que recibe `flatten` no sea un array. Por ejemplo, actualmente `flatten([1, 2])` truena porque dentro del cuerpo de esa función, intentamos hacer `a.concat(b)`, pero `a` en este caso es un `Number` y no tiene una función `concat`. Si empezamos con un `[]`, nos ahorramos el problema. :D

Reply:
Para manejar `[]` anidados en el input de `flatten`, basta con identificar si el siguiente valor en la iteración es un `[]`.

Hay varias maneras de lograr eso, te comparto tres de ellas:

```js
// la mejor
if (Array.isArray([])) {
  console.log("true!")
}

// la que se usaba antes de Array.isArray
if (Object.prototype.toString.call([]) === "[object Array]") {
  console.log("true!")
}

// no tan recomendada
if ([] instanceof Array) {
  console.log("true!")
}
```

Como mencionaste, si el siguiente valor es un `[]`, podemos usar recursivamente nuestra función `flatten`!

```diff
function flatten(nestedArray) {
    return nestedArray.reduce((a, b) => {
+      if (Array.isArray(b)) {
+        return a.concat(flatten(b)) // muy importante concatenar el resultado de la llamada recursiva de flatten
+      } else {
+        return a.concat(b)
+      }
    }, []);
}
```

Reply:
Ohhhh no sabia eso del segundo argumento de `reduce`, que macizo queda al final, got it.

Reply:
También, esto ya es más como un tip, un shortcut mental para las funciones que pasamos a `reduce` puede ser nombrar el primer parámetro de esa función como `(accumulator, value)` donde `value` vendría siendo el item que estés iterando actualmente.
```diff
function flatten(nestedArray) {
+    return nestedArray.reduce((accumulator, value) => {
+      if (Array.isArray(value)) {
+        return accumulator.concat(flatten(value)) // muy importante concatenar el resultado de la llamada recursiva de flatten
+      } else {
+        return accumulator.concat(value)
+      }
    }, []);
}
```

Reply:
Ok, va, si cai en la mala practica de nombrar ambiguamente. Lo cambio de una vez y en los que vea que use arrow functions 👍